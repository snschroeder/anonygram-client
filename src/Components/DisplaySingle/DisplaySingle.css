

.single-image {
  height: 50vh
}

.upvote-wrapper:hover{
  cursor: pointer;
}
.upvote-wrapper {
  position: relative;
  display: flex;
  direction: row;
  justify-content: center;
}

.upvote-count {
  position: absolute;
  bottom: 5px;
  right: 13px;
  font-size: 14px;
  font-weight: bold;
}

.upvote-button {
  position: absolute;
  z-index: auto;
  bottom: 10px;
  right: 30px;
  border: none;
  border-radius: 3px;
  width: 40px;
  height: 50px;
  background: grey;
  opacity: .5;
  color: white;
  font-size: 20px;
  font-weight: bold;
  text-align: center;
}

.upvote-button-no-auth {
  position: absolute;
  z-index: auto;
  bottom: 10px;
  right: 30px;
  border: none;
  border-radius: 3px;
  width: 40px;
  height: 30px;
  background: grey;
  opacity: .5;
  color: white;
  font-size: 20px;
  font-weight: bold;
  text-align: center;
}

.timestamp-display{
  text-align: left;
  color: grey;
  text-transform: uppercase;
  font-size: 10px;
  padding-left: 5px;
  width: 100%;
}


@media screen and (min-width: 1025px){
  .DisplaySingle {
    width: 50%;
    margin: auto;
  }
}
/* scroll bar */
::-webkit-scrollbar {
  width: 20px;
}

::-webkit-scrollbar-track {
  box-shadow: inset 0 0 5px grey; 
  border-radius: 10px;
}

::-webkit-scrollbar-thumb {
  border-radius: 10px;
}


