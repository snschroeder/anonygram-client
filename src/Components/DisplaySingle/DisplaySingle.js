import React, { Component } from 'react';
import moment from 'moment';
import CommentFeed from '../CommentFeed/CommentFeed';
import CommentApi from '../../services/comment-api-service';
import ImageContext from '../../contexts/ImageContext';
import TokenService from '../../services/token-service';
import { KeyboardArrowUp } from '@material-ui/icons';
import './DisplaySingle.css';

export default class DisplaySingle extends Component {
  static contextType = ImageContext; 
  
  state = {
    image: {}, 
    comments: [], 
    loading: false, 
  }

  convertTime = (timestamp) => {
    return moment((new Date(timestamp)).toString()).fromNow();
  }

  componentDidMount = () => {
    // scroll immediately to top-left when component renders
    window.scrollTo(0, 0);

    const submissionId = this.props.submissionId;
    const image = this.context.images.find(img => img.id === Number(submissionId));

    this.setState({ loading: true });
    CommentApi.getComments(submissionId)
      .then((res) => {
        this.setState({
          image,
          comments: res,
          loading: false,
        })
      });
  }

  setCommentsByPush = (comment) => {
    const commentsArr = this.state.comments.map(c => c);
    commentsArr.push(comment);
    this.setState({ comments: commentsArr })
  } 

  render = () => {
    const userLoggedIn = TokenService.hasAuthToken();
    const userIDFromToken = TokenService.parseAuthToken() !== undefined ? TokenService.parseAuthToken().id : ''; 

    //This conditional is needed to keep componentDidMount in CommentFeed from running before the needed props are ready to be passed in (to generate usernames)
    if (this.state.comments == null) {
      return null;
    }
    if (this.state.loading === true) {
      return (
        <div className='loader-container'>
          <div className="loader"></div>
        </div>
      )
    } else {
      const { id, image_url, image_text, create_timestamp, karma_total, user_id } = this.state.image
      return (
        <div className='DisplaySingle'>
          <img className='single-image' src={image_url} alt={image_text}/>
          { /* If user is logged in, render enabled upvote button */
          userLoggedIn && userIDFromToken !== user_id  ? (
            <div className='upvote-wrapper' onClick={() => this.context.incrementUpvotes(id)}>
              <div className="upvote-button">
               <KeyboardArrowUp fontSize="large"/>
               <p className="upvote-count">{karma_total}</p>
              </div>
            </div>
          ) : (
            <div className='upvote-wrapper'>
              <div className="upvote-button-no-auth">
               <p className="upvote-count">{karma_total}</p>
              </div>
            </div>
          )}
          <div className='timestamp-display'>{this.convertTime(create_timestamp)}</div>

          <p>{image_text}</p>
          <CommentFeed id={this.state.image.id} comments={this.state.comments} setCommentsByPush={this.setCommentsByPush} userLoggedIn={userLoggedIn}/>
        </div>
      )
    }
  }
}