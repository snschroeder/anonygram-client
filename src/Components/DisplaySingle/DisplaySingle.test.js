import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import DisplaySingle from './DisplaySingle';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('DisplaySingle component', () => {
  DisplaySingle.contextTypes = {
    images: PropTypes.array,
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter><DisplaySingle /></BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  const props = {
    submissionId: 1,
  };
  const context = {
    images: [
      {
        create_timestamp: new Date('2019-12-10T00:40:55.725Z'),
        id: 1,
        image_url: 'https://via.placeholder.com/150',
        image_text: 'this is image text',
        karma_total: 3,
      },
    ],
    incrementUpvotes: () => {}
  };

  const state = {
    image: {
      create_timestamp: new Date('2019-12-10T00:40:55.725Z'),
      id: 1,
      image_url: 'https://via.placeholder.com/150',
      image_text: 'this is image text',
      karma_total: 3,
    },
    comments: [
      {
        comment_timestamp: new Date('2019-12-10T00:40:55.725Z'),
        user_id: '2d644486-26dc-4ad3-b460-3c6895aa260f',
        comment_id: '2d644486-26dc-4ad3-b460-3c6895aa260f',
        comment_text: 'some comment text',
      },
    ],
  };
  it('renders a submission given props and context', async () => {
    const wrapper = shallow(<DisplaySingle {...props} />, { context });
    const instance = wrapper.instance()
    // for testing snapshot, need to hard code the X days ago
    // to something consistent
    jest
      .spyOn(instance, 'convertTime')
      .mockImplementation(() => '1 day ago')
    wrapper.setState({
      loading: false,
      image: state.image,
      comments: state.comments,
    });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
