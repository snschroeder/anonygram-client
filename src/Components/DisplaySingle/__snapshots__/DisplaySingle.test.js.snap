// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplaySingle component renders a submission given props and context 1`] = `
<div
  className="DisplaySingle"
>
  <img
    alt="this is image text"
    className="single-image"
    src="https://via.placeholder.com/150"
  />
  <div
    className="upvote-wrapper"
  >
    <div
      className="upvote-button"
    >
      <KeyboardArrowUpIcon
        fontSize="large"
      />
      <p
        className="upvote-count"
      >
        3
      </p>
    </div>
  </div>
  <div
    className="timestamp-display"
  >
    1 day ago
  </div>
  <p>
    this is image text
  </p>
  <CommentFeed
    comments={
      Array [
        Object {
          "comment_id": "2d644486-26dc-4ad3-b460-3c6895aa260f",
          "comment_text": "some comment text",
          "comment_timestamp": 2019-12-10T00:40:55.725Z,
          "user_id": "2d644486-26dc-4ad3-b460-3c6895aa260f",
        },
      ]
    }
    id={1}
    setCommentsByPush={[Function]}
    userLoggedIn={false}
  />
</div>
`;
